### ISSUE TITLE
How to override URLs?

### ISSUE THREAD

**Issue by geoffroy-noel-ddh** [CloudCannon/pagefind#686]
I believed I've followed the instructions on how to override page urls but that doesn't seem to work in the results

## Step to reproduce:

Create index.html under a new folder called site:

```html
<!DOCTYPE html>
<html lang="en">
<head>
        <meta charset="utf-8">
        <title>Page title</title>
        <link data-pagefind-meta="url[href]" rel="canonical" href="https://example.com/other-url">
</head>
<body>
 <h1>title 1</h1>
 <p>A simple sentence to be indexed.</p>
</body>
</html>
```

Add a search page under site/search.html:

```html
<html>
<body>
<link href="/pagefind/pagefind-ui.css" rel="stylesheet">
<script src="/pagefind/pagefind-ui.js"></script>
<div id="search"></div>
<script>
    window.addEventListener('DOMContentLoaded', (event) => {
        new PagefindUI({ element: "#search", showSubResults: true });
    });
</script>
</body>
</html>
```

Index and serve with pagefind (I'm using version 1.1.0):

`npx pagefind --site site --serve`

Point your browser to localhost:1414/search.html.

Search for 'sent'

You correctly get one result:

![image](https://github.com/user-attachments/assets/f2174287-cc43-4de6-b357-44e824084b2c)

However that hyperlink leads to http://localhost:1414/ .

Instead, [according to the documentation](https://pagefind.app/docs/ui-usage/#overriding-the-url-of-a-result), I was expecting the URL to be https://example.com/other-url .

Did I misunderstand or forget something or made a mistake? Or is that a bug?

Thanks in advance for the assistance.

**Comment by bglw** [comment]
Hey @geoffroy-noel-ddh 👋 

Thanks for the great report and reproduction. Yes this is indeed a bug! I'll get it fixed up.

**Comment by kylejrp** [comment]
@bglw Wanted to confirm that this bug is still present in `pagefind@1.4.0`. I'm setting a `<link rel="canonical" href="https://example.com/" data-pagefind-meta="url[href]" />` tag in the `<head>` of my page, but I'm still seeing the incorrect URLs in the index. Setting it inside my `<body data-pagefind-body>` with a `<meta data-pagefind-meta="url[content]" content="https://example.com/" />` doesn't seem to work, either.

The raw data JSON looks something like this - it's parsing the override as regular metadata in `meta.url` instead of overriding the root `url`:

```
{
  "url": "/article-title-goes-here/",
  "content": "Article title goes here. Content goes here.",
  ...
  "meta": {
    "date": "2021-04-20T07:00:00.000Z",
    "title": "Article title goes here",
    "url": "https://example.com/",
  },
  ...
}
```

I have some pages on my site that are only present for when the index is built and then are deleted from the final bundle after Pagefind is done with it, so those search result links would be broken on my real site (eg. pages in `/search-index-shadow/external-content/page-title/`, which aren't real pages on my site and only exist to index external content that I've mirrored on my site) 

I can work around it by rewriting the URLs on my search front-end, but it would be nice if the index had the correct URL for things. 😄

**Comment by bglw** [comment]
Ah, specifically the url override is explicitly done through meta, for example the default UI itself implements:

https://github.com/Pagefind/pagefind/blob/04c327a97a128c789dbe9f09e9d06129376c2f8f/pagefind_ui/default/svelte/result.svelte#L39-L44

I'd need to check whether this issue still exists, but it's expected (to some degree) that the `url` is always the Pagefind-generated URL. 

(There's a new ground-up Pagefind UI coming in ~1 week which also uses the same logic)

**Comment by kylejrp** [comment]
Ah! That's not the behavior I expected, but now that I reread the documentation, that is exactly what you say it does. Hmm. 

It's a bit tricky because I want to index copies of external content on my personal site (eg. the transcript of a podcast I was on, a Bluesky post, etc.). My plan was to just generate fake pages so they'd get parsed by Pagefind's parser with all my metadata/filters/sorts and I'd just override the index `url`, but that doesn't seem possible. (unless I work around it by reading and prioritizing the `meta.url` field if it exists)

I could try to index it with `index.addCustomRecord` which I assume lets me set the `url` field directly, but then I have to set up another mapping for all my metadata/filters/sorts just for those records.

I guess I'm nitpicking a bit though 😅 Just feels weird to have to do `data.meta.url ?? data.url` on my end as a consumer of the Pagefind index/search results to account for it in my own custom search components. Also feels weird to ship internal-only broken URLs in the index, though who's really going to see that anyways, haha.

**Comment by bglw** [comment]
Yeah it's kinda just where things landed ha. Nothing defined on the page ever escapes meta, the title is always in meta for example. The url is just a special case since Pagefind always does a lot of processing on it both when indexing and when searching (with baseURL for example, which it will _not_ try append to the meta url). 

I think at the time I also didn't want to add a special meta that reached into other parts of the indexing code. (It _is_ special meta since the UI uses it, but it's not special to pagefind-itself). I agree it would be nice to do something outside of `index.addCustomRecord`, but yes for now the pattern is "consumers can choose whether the tagged url metadata overrides the official url". 

I don't have a great idea off the dome for a way I'd be happy configuring this, but I'm sure such a way exists as a config file option or what have you

**Comment by kylejrp** [comment]
Makes sense, thanks for that context. I agree with your design decision that metadata shouldn't be able to escape into the more complex parts of indexing code.

I think there's room to improve documentation to make this clear for those writing a custom search client using the NodeJS API, though. I have ignored the "Using the Default UI" pages because I'm not using it at all. Some places that could be improved:
- [Using the search API](https://pagefind.app/docs/api/): Could use a "Reading the overridden URL of a result" section that calls out the behavior or `meta.url` (just that it has no special behavior). Could also have a "Reading the overridden title or image of a result" section too.
- [Metadata in the JS API](https://pagefind.app/docs/js-api-metadata/): Could mirror the explanation of the metadata fields in [Metadata in the Default UI](https://pagefind.app/docs/default-ui-metadata/), namely that you should handle the overridden `title`, `image`, `image_alt`, and `url` metadata to match the Default UI's behavior.
- [Indexing content using the NodeJS API](https://pagefind.app/docs/node-api/) and [Search API Reference](https://pagefind.app/docs/api-reference/) could reference these special metadata cases too
- A central definition of the types returned in each result set and if they're escaped or not would also be appreciated - I know there's already type definitions for them when I look in my IDE, and I can only find the one note about the fact that `excerpt` is HTML encoded already, but nothing on the encoding state of the other fields.

If you'd like, I can put a pull request up with some documentation changes! Happy to make the trail better for those that come after me. 😁 Just point me in the right direction if you'd like me to help out.

**Comment by bglw** [comment]
Heya @kylejrp — closing this issue for now in general cleanup, sorry for missing your questions there!

I'm open to any and all docs PRs really, so feel free to tackle it there! I can give any specific feedback on a PR :)
